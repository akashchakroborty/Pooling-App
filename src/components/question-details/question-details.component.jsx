import React, { Fragment, useState, useEffect } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import {
  QuestionDetailsHeader,
  QuestionDetailsContainer,
  QuestionTitle,
  ChoiceContainer,
  ChoiceList,
  SaveButton
} from "./question-details.styles";
import QuestionDetailsActionTypes from "../../redux/question-details/question-details.types";
import { createStructuredSelector } from "reselect";
import {
  selectQuestion,
  selectIsQuestionDetailsFetching
} from "../../redux/question-details/question-details.selector";

import { calculatePercent } from "./question-details.utils";

const QuestionDetails = ({
  dispatch,
  match,
  history,
  selectQuestion,
  isFetching
}) => {
  const { question, choices } = selectQuestion;

  const [selectedChoice, setSelectedChoice] = useState({
    title: "",
    url: null
  });

  useEffect(() => {
    const { questionId } = match.params;
    dispatch({
      type: QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_START,
      payload: questionId
    });
  }, [dispatch, match.params]);

  const handleChoiceOnClick = ({ choice, url }) => {
    setSelectedChoice({ ...selectedChoice, title: choice, url });
  };

  const handleSubmit = event => {
    event.preventDefault();
    dispatch({
      type: QuestionDetailsActionTypes.PLACE_VOTE_START,
      payload: {
        url: selectedChoice.url,
        callback: () => {
          history.push(`${match.url}/voteSuccess`);
        }
      }
    });
  };

  return (
    <div>
      <QuestionDetailsHeader>
        <h1>Question Details</h1>
      </QuestionDetailsHeader>
      <QuestionDetailsContainer>
        {isFetching ? (
          <h1>Loading...</h1>
        ) : (
          <Fragment>
            <QuestionTitle>Question: {question}</QuestionTitle>
            <ChoiceContainer onSubmit={handleSubmit}>
              {choices &&
                choices.map((choice, index) => (
                  <ChoiceList
                    key={index}
                    className={`${
                      selectedChoice.title === choice.choice ? " active" : ""
                    }`}
                    onClick={event => {
                      event.preventDefault();
                      handleChoiceOnClick(choice);
                    }}
                  >
                    <span>{choice.choice}</span>
                    <span>{choice.votes}</span>
                    <span className="text-center">
                      {calculatePercent(choice.votes, choices)} %
                    </span>
                    <progress
                      value={Math.round(
                        calculatePercent(choice.votes, choices)
                      )}
                      max="100"
                    ></progress>
                  </ChoiceList>
                ))}
              <SaveButton type="submit" disabled={!!!selectedChoice.url}>
                Save vote
              </SaveButton>
            </ChoiceContainer>
          </Fragment>
        )}
      </QuestionDetailsContainer>
    </div>
  );
};

QuestionDetails.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  selectQuestion: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired
};

const mapStateToProps = createStructuredSelector({
  selectQuestion: selectQuestion,
  isFetching: selectIsQuestionDetailsFetching
});

export default connect(mapStateToProps)(QuestionDetails);
