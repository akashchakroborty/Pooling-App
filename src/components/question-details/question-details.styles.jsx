import styled from "styled-components";

export const QuestionDetailsHeader = styled.div`
  text-align: center;
  background: #0d223f;
  max-width: 80vw;
  width: 80%;
  margin: 0 auto;
  border-top: 10px solid #ffc627;
  height: 100px;
  padding-top: 12px;
`;

export const QuestionDetailsContainer = styled.div`
  max-width: 80vw;
  width: 80%;
  margin: 0 auto;
  margin: 1rem auto;
`;

export const QuestionTitle = styled.div`
  color: #fff;
  border-radius: 5px;
  font-size: 150%;
  margin: 1rem auto;
  background-color: #0d223f;
  padding: 20px;
  color: #ffc627;
`;

export const ChoiceContainer = styled.form`
  color: #fff;
  font-size: 150%;
  grid-column: col / span 4;
  grid-row: row 2;
`;

export const ChoiceList = styled.div`
  background: #ffc627;
  height: 70px;
  color: black;
  margin-bottom: 2px;
  transition: border-radius 0.1s ease-in-out 0s,
    letter-spacing 0.1s ease-in-out 0s;
  padding: 0 0.5rem;
  text-align: center;
  align-items: center;
  border-radius: 5px;
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  &.active {
    opacity: 0.7;
  }

  &:hover {
    opacity: 0.8;
    cursor: pointer;
  }
  progress {
    width: auto;
    background: black;
  }
`;

export const SaveButton = styled.button`
  float: right;
  margin-top: 1rem;
`;
