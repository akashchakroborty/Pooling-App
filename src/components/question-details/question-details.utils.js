/**
 * Based on the Qty of votes calculate the percent
 * @param {number} votes
 * @param {number} totalvotes
 * @return {string} percevalue
 */
export const calculatePercent = (votes, choiseArray = []) => {
  let totalvotes = (function(choiseArray) {
    let sum = 0;
    for (let i = 0; i < choiseArray.length; i++) {
      sum += choiseArray[i].votes;
    }
    return sum;
  })(choiseArray);

  const percevalue = isNaN(
    parseInt((parseInt(votes, 10) * 100) / totalvotes, 10).toFixed(2)
  )
    ? 0
    : ((parseInt(votes, 10) * 100) / totalvotes).toFixed(2);
  return percevalue;
};
