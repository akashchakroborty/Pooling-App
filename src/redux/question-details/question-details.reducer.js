import QuestionDetailsActionTypes from "./question-details.types";
const INITIAL_STATE = {
  question: {},
  isFetching: false,
  errorMessage: undefined
};

const questionDetailsReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_START:
      return {
        ...state,
        isFetching: true
      };
    case QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_SUCCESS:
      return {
        ...state,
        isFetching: false,
        question: action.payload
      };
    case QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_FAILURE:
      return {
        ...state,
        isFetching: false,
        errorMessage: action.payload
      };
    default:
      return state;
  }
};

export default questionDetailsReducer;
