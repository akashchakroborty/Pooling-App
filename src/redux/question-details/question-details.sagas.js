import { call, put, takeLatest, all } from "redux-saga/effects";
import QuestionDetailsActionTypes from "./question-details.types";
import axios from "axios";

const ROOT_URL = "https://polls.apiblueprint.org";

function poolQuestionDetailsApi(id) {
  return axios.request({
    method: "get",
    url: `${ROOT_URL}/questions/${id}`
  });
}

export function* fetchQuestionDetailsAsync(action) {
  try {
    let { data } = yield call(poolQuestionDetailsApi, action.payload);
    yield put({
      type: QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_SUCCESS,
      payload: data
    });
  } catch (error) {
    yield put({
      type: QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_FAILURE,
      payload: error.message
    });
  }
}
export function* fetchQuestionDetailsStart() {
  yield takeLatest(
    QuestionDetailsActionTypes.FETCH_QUESTIONDETAILS_START,
    fetchQuestionDetailsAsync
  );
}

export function* questionDetailsSagas() {
  yield all([call(fetchQuestionDetailsStart)]);
}
