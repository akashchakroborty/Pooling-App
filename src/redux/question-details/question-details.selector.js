import { createSelector } from "reselect";

const selectquestionDetails = state => state.questionDetails;

export const selectQuestion = createSelector(
  [selectquestionDetails],
  questionDetails => questionDetails.question
);

export const selectIsQuestionDetailsFetching = createSelector(
  [selectquestionDetails],
  questionDetails => questionDetails.isFetching
);
