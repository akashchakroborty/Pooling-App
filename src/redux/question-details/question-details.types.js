const QuestionDetailsActionTypes = {
  FETCH_QUESTIONDETAILS_START: "FETCH_QUESTIONDETAILS_START",
  FETCH_QUESTIONDETAILS_SUCCESS: "FETCH_QUESTIONDETAILS_SUCCESS",
  FETCH_QUESTIONDETAILS_FAILURE: "FETCH_QUESTIONDETAILS_FAILURE"
};

export default QuestionDetailsActionTypes;
