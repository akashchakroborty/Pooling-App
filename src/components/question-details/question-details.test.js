import React from "react";
import { shallow } from "enzyme";
import { QuestionDetails } from "./question-details.component";
import { ChoiceList, ChoiceContainer } from "./question-details.styles";

describe("QuestionDetails component", () => {
  let mockDispatch;
  let mockSelectQuestion;
  let mockIsFetching;
  let mockHistory;
  let mockMatch;
  let wrapper;

  beforeEach(() => {
    mockDispatch = jest.fn();
    mockHistory = {
      push: jest.fn()
    };
    mockMatch = {
      params: {
        questionId: 1
      }
    };
    mockSelectQuestion = {
      question: "mock",
      choices: [
        {
          choice: "mock",
          votes: 1,
          url: "mock"
        }
      ]
    };
    mockIsFetching = false;

    const mockProps = {
      dispatch: mockDispatch,
      history: mockHistory,
      match: mockMatch,
      selectQuestion: mockSelectQuestion,
      isFetching: mockIsFetching
    };

    wrapper = shallow(<QuestionDetails {...mockProps} />);
  });

  it("should render QuestionDetails component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render an equal number of ChoiceList as the Selected questions choices prop", () => {
    expect(wrapper.find(ChoiceList).length).toEqual(
      mockSelectQuestion.choices.length
    );
  });

  it("should render Loading if isFetching is true", () => {
    const mockProps = {
      dispatch: mockDispatch,
      history: mockHistory,
      match: mockMatch,
      selectQuestion: {},
      isFetching: true
    };

    const newWrapper = shallow(<QuestionDetails {...mockProps} />);
    expect(newWrapper.exists(".loading")).toBe(true);
  });

  it("should call handleSubmit when ChoiceContainer is Submited", () => {
    const event = {
      preventDefault() {}
    };
    wrapper.find(ChoiceContainer).simulate("submit", event);
    expect.any(Function);
  });

  it("should call handleChoiceOnClick when ChoiceList is clicked", () => {
    const event = {
      preventDefault() {}
    };
    wrapper.find(ChoiceList).simulate("click", event);
    expect.any(Function);
  });
});
