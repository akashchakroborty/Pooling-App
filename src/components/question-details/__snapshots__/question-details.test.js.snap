// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetails component should render QuestionDetails component 1`] = `ShallowWrapper {}`;
